src/screenshot/screenshot.c

```diff
@@ -148,9 +149,149 @@ static int method_screenshot(sd_bus_message *msg, void *data,
 	return 0;
 }
 
+static bool spawn_chooser(int chooser_out[2]) {
+	pid_t pid = fork();
+	if (pid < 0) {
+		perror("fork");
+		return false;
+	} else if (pid == 0) {
+		close(chooser_out[0]);
+
+		dup2(chooser_out[1], STDOUT_FILENO);
+		close(chooser_out[1]);
+
+		execl("/bin/sh", "/bin/sh", "-c", "grim -g \"$(slurp -p)\" -t ppm -", NULL);
+
+		perror("execl");
+		_exit(127);
+	}
+
+	int stat;
+	if (waitpid(pid, &stat, 0) < 0) {
+		perror("waitpid");
+		return false;
+	}
+
+	close(chooser_out[1]);
+	return stat == 0;
+}
+
+static struct xdpw_ppm_pixel *exec_color_picker() {
+	int chooser_out[2];
+	if (pipe(chooser_out) == -1) {
+		perror("pipe chooser_out");
+		return NULL;
+	}
+
+	if (!spawn_chooser(chooser_out)) {
+		logprint(ERROR, "Selection failed");
+		close(chooser_out[0]);
+		return NULL;
+	}
+
+	FILE *f = fdopen(chooser_out[0], "r");
+	if (f == NULL) {
+		perror("fopen pipe chooser_out");
+		close(chooser_out[0]);
+		return NULL;
+	}
+
+	struct xdpw_ppm_pixel *img = malloc(sizeof(struct xdpw_ppm_pixel));
```

Comment:
To simplify the code, we can take a `struct xdpw_ppm_pixel *` as an argument of this function, and fill its fields.

Reply:
Just to clarify, then the code calling this function would look like this?
```
struct xdpw_ppm_pixel *img = malloc(sizeof(struct xdpw_ppm_pixel));
if (!exec_color_picker(img)) {
	free(img);
	return -1;
}
```

Reply:
No need to allocate on the heap:

```c
struct xdpw_ppm_pixel img = {0};
if (!exec_color_picker(&img)) {
    return -1;
}
```